File: huipyo/src/week04/pgs42839/Solution.js
function solution(numbers) {
    var stack = [];
    let primeNumbers = new Set();
    let numberList = numbers.split("");
    function getNumbers() {
        for (let i = 0; i < numberList.length; i++) {
            stack.push(i);
            if (stack.length !== 0) {
                let n = parseInt([...new Set(stack)].map(index => numberList[index]).join(""));
                if (isPrimeNumber(n)) primeNumbers.add(n);
                if (stack.length - 1 < numberList.length - 1) getNumbers();
            }
            stack.pop();
        }
    }

    getNumbers();
    return primeNumbers.size;
}

function isPrimeNumber(n) {
    if (n <= 1) return false;
    if (n <= 3) return true;
    if (n % 2 === 0 || n % 3 === 0) return false;
    let i = 5;

    while (i * i <= n) {
        if (n % i === 0 || n % (i + 2) === 0) return false;
        i += 6;
    }
    return true;
}

File: huipyo/src/week04/pgs42839/Solution.java
package week04.pgs42839;

import java.math.BigInteger;
import java.util.*;
import java.util.stream.*;

class Solution {
    public int solution(String numbers) {
        LinkedList<Integer> stack = new LinkedList<>();
        Set<Integer> primeNumbers = new HashSet<>();
        List<String> numberList = Arrays.asList(numbers.split(""));
        Runnable getNumbers = new Runnable() {
            @Override
            public void run() {
                for (int i = 0; i < numberList.size(); i++) {
                    stack.push(i);
                    if (!stack.isEmpty()) {
                        int n = Integer.parseInt(stack.stream().distinct().map(numberList::get).collect(Collectors.joining()));

                        if (BigInteger.valueOf(n).isProbablePrime(10)) primeNumbers.add(n);
                        if (stack.size() - 1 < numberList.size() - 1) run();
                    }
                    stack.pop();
                }
            }
        };

        getNumbers.run();
        return primeNumbers.size();
    }
}